tools/fuzzy_proof_generator_tool/test/FuzzyProofGenerator/assertPatternValidity.test.ts

```diff
@@ -0,0 +1,189 @@
+// Copyright 2019 OpenST Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// ----------------------------------------------------------------------------
+
+import 'mocha';
+import { assert } from 'chai';
+import FuzzyProofGenerator from '../../src/FuzzyProofGenerator';
+
+describe('FuzzyProofGenerator::assertPatternValidity', (): void => {
+  it('Reverts if a pattern is empty.', (): void => {
+    assert.throws(
+      (): void => FuzzyProofGenerator.assertPatternValidity(''),
+    );
+  });
+
+  it('Reverts if a pattern ends with an extension node.', (): void => {
+    assert.throws(
+      (): void => FuzzyProofGenerator.assertPatternValidity('e'),
+    );
+
+    assert.throws(
+      (): void => FuzzyProofGenerator.assertPatternValidity('be'),
+    );
+  });
+
+  it('Reverts if a pattern contains two consecutive extension nodes.', (): void => {
+    assert.throws(
+      (): void => FuzzyProofGenerator.assertPatternValidity('eel'),
+    );
+
+    assert.throws(
+      (): void => FuzzyProofGenerator.assertPatternValidity('beel'),
+    );
+  });
+
+  it('Reverts if a leaf node follows an extension nodes.', (): void => {
+    assert.throws(
+      (): void => FuzzyProofGenerator.assertPatternValidity('el'),
+    );
+
+    assert.throws(
+      (): void => FuzzyProofGenerator.assertPatternValidity('bel'),
+    );
+  });
+
+  it('Reverts if a leaf node is not the last one in pattern.', (): void => {
+    assert.throws(
+      (): void => FuzzyProofGenerator.assertPatternValidity('lbb'),
+    );
+
+    assert.throws(
+      (): void => FuzzyProofGenerator.assertPatternValidity('blbb'),
+    );
+  });
+
+  it('Reverts if a pattern does not end with two branch nodes.', (): void => {
```

Comment:
This I still don't understand. What about the following example:

Paths that exist

| path | value |
| --- | --- |
| `aa` | `0` |
| `aab` | `1` |
| `aac` | `2` |

Root would be an extension node like this, right? Random "next hash".

| path | next |
| --- | --- |
| `aa` | `0x2934` |

And then node `0x29349247` would be a branch node like this, right? Random "next hashes".

| `0` | `1` | ... | `b` | `c` | ... | `f` | value |
|---|---|---|---|---|---|---|---|
| | | | `0xbbbb` | `0xcccc` | | | `0` |

Now would this not be valid? It ends on a single branch node.
| pattern | path | value |
| --- | --- | --- |
| `eb` | `aa` | `0` |

---

Also, from my understanding, the whole discussion is pointless in the context of our ethereum proofs, as the last node will always be a leaf node as the path length is the same for all values (20 bytes), right?

Reply:
No. Extension nodes don't have a value. That's exactly what the value field on the branch node is for. Or am I completely confusing this?

![YZGxe](https://user-images.githubusercontent.com/1196244/57921173-09230780-789d-11e9-9df2-f8d7e5dd95ab.png)

Reply:
True. Extension node has no value. However, a verification library we are using checks that if 2-elements array (extension or leaf) ends pattern, it attempts to compare against its value. In this case "value" will be extension node "key" segment.

Reply:
Ah ok so the library is "broken" 😖

Reply:
It's the way how the library verifies. For example, if the handling for branch node would be similar to leaf/extension nodes, then the pattern "eb" for example would be valid.

Reply:
ok 👍 
as discussed the library should not be changed as part of this PR.